jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdAuthenticator.java

```diff
@@ -494,7 +479,80 @@ protected String getChallengeUri(HttpServletRequest request)
     @Override
     public boolean secureResponse(ServletRequest req, ServletResponse res, boolean mandatory, User validatedUser)
     {
-        return true;
+        return req.isSecure();
+    }
+
+    private UriRedirectInfo removeAndClearCsrfMap(HttpSession session, String csrf)
+    {
+        @SuppressWarnings("unchecked")
+        Map<String, UriRedirectInfo> csrfMap = (Map<String, UriRedirectInfo>)session.getAttribute(CSRF_MAP);
+        if (csrfMap == null)
+            return null;
+
+        UriRedirectInfo uriRedirectInfo = csrfMap.remove(csrf);
+        csrfMap.clear();
+        return uriRedirectInfo;
+    }
+
+    private Map<String, UriRedirectInfo> ensureCsrfMap(HttpSession session)
+    {
+        @SuppressWarnings("unchecked")
+        Map<String, UriRedirectInfo> csrfMap = (Map<String, UriRedirectInfo>)session.getAttribute(CSRF_MAP);
+        if (csrfMap == null)
+        {
+            // Create a custom Map so we can only have a limited number of request URIs saved.
+            csrfMap = new LinkedHashMap<>()
+            {
+                private static final int MAX_SIZE = 64;
+
+                @Override
+                protected boolean removeEldestEntry(Map.Entry<String, UriRedirectInfo> eldest)
+                {
+                    return size() > MAX_SIZE;
+                }
+            };
+            session.setAttribute(CSRF_MAP, csrfMap);
+        }
+        return csrfMap;
+    }
+
+    private static class UriRedirectInfo
```

Comment:
@janbartel are there strict rules about custom classes saved to the `HttpSession`? Like class should be `public`, `Serializable`, etc.

Reply:
The best thing to do is to make any attribute you store in the `HttpSession` `be Serializable`. Otherwise your attribute could implement the `HttpSessionActivationListener` interface instead: it can transform itself in some way on the `sessionWillPassivate()` and `sessionWillActivate()` methods.